### ISSUE TITLE
Adapt plugin-theme to support the new scale setting

### ISSUE THREAD

**Issue by joostdecock** [freesewing/freesewing#1638]
**Problem description**

We've added improved support for very small (and large) patterns to our roadmap. 
Think patterns for dolls/giants.

The problem is that while the patterns themselves are made-to-measure, some things don't scale:

 - Stroke-width
 - Arrow sizes (also based on stroke width)
 - Font sizes
 - Snippets like buttons and scalebox and whatnot

Here's an example that illustrates the problem:
![Screenshot 2022-01-11 at 17-15-36 aaron](https://user-images.githubusercontent.com/1708494/148980015-783566dd-c994-4f3d-ae70-6ebf0d3f6d29.png)

**What is it that you would like to see happen?**

In our frontend, we handle this with CSS variables. Alas, you cannot set CSS variables in SVG (or if you can, then I haven't figured out how). So for stand-along SVG files this is not a solution.

So, we should adapt our `plugin-theme` package to take this variable on board.
Specifically, rather than add a bunch of static CSS, it should generate the CSS dynamically, and multiply font sizes, stroke widths and so on with the `settings.scale` value (which defaults to 1).

Currently, the plugin imports the CSS, then adds it to the pattern (there's also a prebuild step that turns SASS into CSS but we don't really use SASS so it should not be a problem to remove it).

Instead, we should have a method that generates the CSS and takes the scale factor as input.

**Comment by joostdecock** [comment]
A bit more info now that I have some more time:

- The `plugin-theme` plugin is a so-called run-time plugin. See: https://freesewing.dev/guides/plugins/types-of-plugins
- I hooks into the `preRender` lifecycle event and received the SVG object. See: https://freesewing.dev/reference/api/hooks/prerender
- The SVG object holds a reference to the full pattern object, precisely for plugins who need access to that. See: https://freesewing.dev/reference/api/svg/pattern
- The settings are stored in the `settings` property of the pattern object
- Which means that your plugin can access the `scale` setting as `svg.pattern.settings.scale`

Then, it's a matter of taking the CSS and adapting stroke-widths, text-size and so on based on the scale. Quick example:

```js
import {version, name} from '../package.json'

// Helper method to avoid having too many digits after the comma
const round = value => Math.round(value * 1e2) / 1e2

const svg = (scale=1) => `
stroke-xs {
  stroke-width: ${round(0.1*scale)};
}
stroke-sm {
  stroke-width: ${round(0.2*scale)};
}
stroke-lg {
  stroke-width: ${round(0.6*scale)};
}
/* and so on... */
`

export default {
  name: name,
  version: version,
  hooks: {
    preRender: svg => {
      // Don't add style twice, because a pattern van get rendered multiple times
      if (svg.attributes.get('freesewing:plugin-theme') === false) {
        svg.style += css(svg.pattern.settings.scale)
        // More stuff here
      }
    }
  }
```

This is from the top of my head, but I hope it makes some sense :)

**Comment by nicholasdower** [comment]
I believe font sizes and stroke widths (and arrows) are now working. Going to tackle the scale box next. That one seems the most interesting.

**Comment by joostdecock** [comment]
Couple of things to help move this along:

- Core now supports the scale setting, which means that:
  - It will apply it to line-height for multi-line text (1)
  - It will automatically scale snippets (logo, notches and so on) based on the scale setting
  - scale is now also available from `Part.shorthand()`
- Components have been updated, Workbench and Draft both have scale support. This means that the development environment for patterns now has `scale` under the advanced settings
- Added `setting.scale` to `i18n` (translation)
- Added CSS variable-based sizing for font-sizes and stroke-widths (1)

(1) By know I guess you will have realized that the CSS for the exported SVG/PDF and the CSS for the development environment/website come from a different place. What I've been doing here is the dev env/site. Whereas plugin-theme provides the embedded SVG/PDF styles.
Just wanted to clarify that I'm helping with some of the other things, but staying out of your lane ;)

**Comment by joostdecock** [comment]
Relevant: #1648

**Comment by joostdecock** [comment]
Note that #1648 is now fully implemented, so it should be easy to test this now.

I've left scalebox untouched, but the other plugins have been adapted.
I'm going to start integrating this new `scale` setting in the frontend. When that's done, I'd like to put out v2.20 for that.

If you'd like to do scalebox, I can wait with that, or I can take it over if you want. Your call.

**Comment by nicholasdower** [comment]
> Note that #1648 is now fully implemented, so it should be easy to test this now.
> 
> I've left scalebox untouched, but the other plugins have been adapted. I'm going to start integrating this new `scale` setting in the frontend. When that's done, I'd like to put out v2.20 for that.
> 
> If you'd like to do scalebox, I can wait with that, or I can take it over if you want. Your call.

I'd love to still work on it but understand if you are busy. I am working on it now in fact but don't have any estimate yet for how long it will take. 

I just spun up plugintest for the first time but haven't figured it out yet. I also can't find the scale setting under advanced but you said it is available in patterns so maybe I should try a pattern next. Let me know if you need to take this over. Otherwise, see you in the PR.

**Comment by nicholasdower** [comment]
> > Note that #1648 is now fully implemented, so it should be easy to test this now.
> > I've left scalebox untouched, but the other plugins have been adapted. I'm going to start integrating this new `scale` setting in the frontend. When that's done, I'd like to put out v2.20 for that.
> > If you'd like to do scalebox, I can wait with that, or I can take it over if you want. Your call.
> 
> I'd love to still work on it but understand if you are busy. I am working on it now in fact but don't have any estimate yet for how long it will take.
> 
> I just spun up plugintest for the first time but haven't figured it out yet. I also can't find the scale setting under advanced but you said it is available in patterns so maybe I should try a pattern next. Let me know if you need to take this over. Otherwise, see you in the PR.

Regarding the scale setting in advanced, maybe I just haven't rebuilt properly. I see it in the code but not in the plugintest UI.

Side note: I believe I would benefit from a walkthrough from someone on how they develop. My workflow feels way too slow to be normal. I am running `yarn build`, `yarn upgrade` and `yarn start` after every change. It takes minutes and it doesn't always work.

**Comment by joostdecock** [comment]
> I'd love to still work on it but understand if you are busy. I am working on it now in fact but don't have any estimate yet for how long it will take.

No stress, it's not on fire. We can wait.

> My workflow feels way too slow to be normal.

Let me try to explain the current state of things, and perhaps it can clarify some things.

**Two levels of dependencies**

There's essentially two levels of dependencies in the monorepo. 
- There's the dependencies for each package under `packages`. Those are configured in the `config/dependencies.yaml` file, and they are probably not causing any issue.
- Then there's the dependencies in the `example` folder of every pattern. That's a similar setup to what `npx create-freesewing-pattern` creates, which is a CRA setup (CRA = create-react-app) with its own benefits and drawbacks. The main benefit being that it works, the main drawback being that it's not flexible.

**The limitations of CRA**

The dependencies under the `example` folder are installed as packages from NPM. Not from the monerepo.
That means that:
 - You need to run `yarn install` there before you can run the dev env the first time, event after the monorepo has installed all dependencies
 - Changes to code in the monorepo will not auto-update
 - CRA does not allow loading dependencies out of the project root (the `example` folder). Even for the pattern code itself, we need to work around it by symlinking it in, which is a bit of a hack.
 
**Forcing CRA updates**

If you updated local package X you can still force an update in the development environment by (using and example where `core` is the dependency we updated and `plugintest` is the pattern we're updating):

```bash
cd packages/core
yarn build
cp dist/* ../plugintest/example/node_modules/@freesewing/core/dist/
cd ../plugintest/example
yarn start
```

Next core update, run:

```bash
cd - && yarn build && cp dist/* ../plugintest/example/node_modules/@freesewing/core/dist/ && cd -
```

Then restart the development environment 

It's a little clunky, but it should work in seconds, rather than minutes.

**What does plugintest bring to the table?**

There's a few reasons why plugintest is going to make your life better:

- It loads all plugin code from the monorepo, not from NPM packages. So updating & rebuilding the plugin code will get picked up automatically
- It does not use plugin-bundle, but rather imports every plugin on its own, so no need to build twice
- It contains a bunch of options (for those plugins where that makes sense) that allow you to control the behavior of the plugin
- There's an option to only show one particular plugin (update `config/index.js` to set the default so it's always the one you're working on

**Why do I not see the scale option in the dev env?**

Recap: Code imported by the pattern will auto-update. Code imported in the development environment is installed from the latest version on NPM and will not auto-update.

The entire development environment is shipped as a React component called `Workbench`. You need to push the updated version in  your dev env like this:

```bash
cd packages/components
yarn build
cp Workbench/* ../plugintest/example/node_modules/@freesewing/components/Workbench/
```
Then restart your dev env. Good news: You only need to do this once (unless you change Workbench itself but you probably won't do that as we're trying to get rid of it which brings me to:

**How can we do better?**

The current development environment is very much geared towards people running `npx create-freesewing-pattern` and it's not optimal for people working in (a fork of) the monorepo directly. I plan to change that by keeping create-freesewing-pattern around for those who just want to develop a pattern, but switching to a NextJS based development environment in the monorepo.

Since we're building freesewing.org (and freesewing.dev) in the monorepo based on NextJS, it should be easy to re-use code for the development environment. Furthermore, NextJS does not have the bolted-down limitations CRA has on importing from outside the package root. It's easy to setup a webpack alias and now we can import all code from the monorepo, rather than from a package deep down in some node_modules folder somewhere.

I'd also like to get rid of the double-level of dependencies by having another package.json and node_modules in `example` and instead just have the development environment live next to the pattern code. We'd just keep it out of the bundle when packaging it, (and make sure to list things as dev dependencies).

This would simplify things, it would allow us to iterate faster, and I hope it will be a significant improvement to the developer experience.

I was planning to work on freesewing.org, but if you'd like I could do it more in parallel, and try to build this new development environment as well. There will be a bunch of shared code anyway, so handling them more or less in parallel is not a big problem.



Last but not least:

> I believe I would benefit from a walkthrough from someone on how they develop.

If this was an opening for a call/chat on the matter, I did not mean to shut it down. It's just that we're in different time zones so async is probably easier. But I'm happy to discuss it if you'd like.

There's also our contributor call this weekend where you are free to join: https://github.com/freesewing/freesewing/discussions/1687

**Comment by nicholasdower** [comment]
#1692

**Comment by nicholasdower** [comment]
Hey @joostdecock, is there more to do on this ticket?

**Comment by nicholasdower** [comment]
> No. You should close it :) I'd do it, but it's your issue so you've earned closing it 😉

Woohoo!
